ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionsFlatStreamReader.scala

```diff
@@ -0,0 +1,243 @@
+    globalPayloadQueriesLimiter: ConcurrencyLimiter,
+    dbDispatcher: DbDispatcher,
+    queryNonPruned: QueryNonPruned,
+    eventStorageBackend: EventStorageBackend,
+    lfValueTranslation: LfValueTranslation,
+    metrics: Metrics,
+)(implicit executionContext: ExecutionContext) {
+  import TransactionsReader._
+  import config._
+
+  private val logger = ContextualizedLogger.get(getClass)
+  private val dbMetrics = metrics.daml.index.db
+
+  private val orderBySequentialEventId =
+    Ordering.by[EventStorageBackend.Entry[Raw.FlatEvent], Long](_.eventSequentialId)
+
+  def streamFlatTransactions(
+      queryRange: EventsRange[(Offset, Long)],
+      filteringConstraints: TemplatePartiesFilter,
+      eventProjectionProperties: EventProjectionProperties,
+  )(implicit loggingContext: LoggingContext): Source[(Offset, GetTransactionsResponse), NotUsed] = {
+    val startExclusiveOffset: Offset = queryRange.startExclusive._1
+    val endInclusiveOffset: Offset = queryRange.endInclusive._1
+    val span =
+      Telemetry.Transactions.createSpan(startExclusiveOffset, endInclusiveOffset)(
+        qualifiedNameOfCurrentFunc
+      )
+    logger.debug(
+      s"streamFlatTransactions($startExclusiveOffset, $endInclusiveOffset, $filteringConstraints, $eventProjectionProperties)"
+    )
+    val sourceOfFlatTransactions: Source[(Offset, GetTransactionsResponse), NotUsed] =
+      doStreamFlatTransactions(
+        queryRange,
+        filteringConstraints,
+        eventProjectionProperties,
+      )
+    sourceOfFlatTransactions
+      .wireTap(_ match {
+        case (_, getTransactionsResponse) =>
+          getTransactionsResponse.transactions.foreach { transaction =>
+            val event =
+              telemetry.Event("transaction", TraceIdentifiers.fromTransaction(transaction))
+            Spans.addEventToSpan(event, span)
+          }
+      })
+      .watchTermination()(endSpanOnTermination(span))
+  }
+
+  private def doStreamFlatTransactions(
+      queryRange: EventsRange[(Offset, Long)],
+      filteringConstraints: TemplatePartiesFilter,
+      eventProjectionProperties: EventProjectionProperties,
+  )(implicit loggingContext: LoggingContext): Source[(Offset, GetTransactionsResponse), NotUsed] = {
+    val (startExclusiveOffset, startExclusiveEventSequentialId): (Offset, Long) =
+      queryRange.startExclusive
+    val (endInclusiveOffset, endInclusiveEventSequentialId): (Offset, Long) =
+      queryRange.endInclusive
+    val createEventIdQueriesLimiter =
+      new QueueBasedConcurrencyLimiter(maxParallelIdCreateQueries, executionContext)
+    val consumingEventIdQueriesLimiter =
+      new QueueBasedConcurrencyLimiter(maxParallelIdConsumingQueries, executionContext)
+    val payloadQueriesLimiter =
+      new QueueBasedConcurrencyLimiter(maxParallelPayloadQueries, executionContext)
+    val decomposedFilters = FilterTableACSReader.makeSimpleFilters(filteringConstraints).toVector
```

Comment:
We need a TODO in `FilterTableACSReader` to extract all the static methods to a separate utility

Reply:
Added